source/agora/consensus/protocol/Nominator.d

```diff
@@ -1231,6 +1241,38 @@ extern(D):
 
         return res;
     }
+
+    import scpd.types.Stellar_types : StellarHash = Hash;
+
+    // `getHashOf` computes the hash for the given vector of byte vector
+    override StellarHash getHashOf (ref vector!Value vals) const nothrow
+    {
+        import libsodium.crypto_generichash;
+        Hash hash = void;
+        crypto_generichash_state state;
+
+        auto dg = () @trusted {
+            crypto_generichash_init(&state, null, 0, Hash.sizeof);
+            scope HashDg dg = (in ubyte[] data) @trusted {
+                crypto_generichash_update(&state, data.ptr, data.length);
+            };
+            return dg;
+        }();
+
+        foreach (value; vals)
+        {
+            const ubyte[] bytes = value[];
+            hashPart(bytes, dg);
+        }
+        void trusted () @trusted
+        {
+            crypto_generichash_final(&state, hash[].ptr, Hash.sizeof);
+        }
+        trusted();
+
+        ubyte[64] bytes = hash[][0..64];
+        return StellarHash(bytes);
+    }
```

Comment:
I think we can improve this but I need to look at it with fresh eyes

Reply:
It would be nice if the hashing routines had support for ranges. I want to be able to do something like this:

```D
const hash = hashFull(vals[].map!(val => val[]));
```

Reply:
I wonder if I just have to flatten the range.. 🤔

Reply:
There is a very strange error when passing ranges to `hashFull`:

```
submodules/crypto/source/agora/crypto/Hash.d(177,22): Error: expression `*record` is `void` and has no value
```

But this part of code is:

```
// Pointers are handled as arrays, only their size must be 0 or 1
else static if (is(T : E*, E))
{
    if (record is null)
        hashVarInt(uint(0), hasher);
    else
    {
        hashVarInt(uint(1), hasher);
        hashPart(*record, hasher);
    }
}
```

How did a range got converted into a pointer? Could it be something to do with `in`?

Reply:
Anyway I think we can add `isInputRange` as a special case in `hashPart` and it could work?

Reply:
Hmm not really possible with input ranges because we only know the length after exhaustion of the input range, but the hash routine should hash the length first.